File: tools/temp_validate_locations.js
const fs = require('fs');
const path = require('path');
const vm = require('vm');

function loadData(filename, variableName) {
  const src = fs.readFileSync(path.join(__dirname, '..', 'data', filename), 'utf-8');
  const sandbox = { exports: {} };
  vm.createContext(sandbox);
  vm.runInContext(src + `\nexports.${variableName} = ${variableName};`, sandbox, { filename });
  return sandbox.exports[variableName];
}

const geojson = JSON.parse(fs.readFileSync(path.join(__dirname, '..', 'data', 'meycauayan-barangays.geojson'), 'utf-8'));
const facilities = loadData('facilities.js', 'FACILITIES_DATA');
const polygons = new Map(geojson.features.map(f => [f.properties.barangay, f.geometry.coordinates[0]]));
const allFacilities = [].concat(...Object.values(facilities).filter(Array.isArray));

function pointInPoly(x, y, poly) {
  let inside = false;
  for (let i = 0, j = poly.length - 1; i < poly.length; j = i++) {
    const xi = poly[i][0], yi = poly[i][1];
    const xj = poly[j][0], yj = poly[j][1];
    const intersect = ((yi > y) !== (yj > y)) &&
      (x < (xj - xi) * (y - yi) / (yj - yi) + xi);
    if (intersect) inside = !inside;
  }
  return inside;
}

const errors = [];
allFacilities.forEach(f => {
  const bar = f.barangay;
  if (!bar) return;
  const poly = polygons.get(bar);
  if (!poly) {
    errors.push(`${f.id} ${f.name} => missing barangay ${bar}`);
    return;
  }
  if (!pointInPoly(f.lng, f.lat, poly)) {
    errors.push(`${f.id} ${f.name} (${bar}) is outside its barangay poly at ${f.lat},${f.lng}`);
  }
});

if (errors.length === 0) {
  console.log('OK');
} else {
  console.log(errors.join('\n'));
}

File: tools/temp_validate_locations.py
import json, re
from pathlib import Path

def parse_js_array(path):
    text = Path(path).read_text(encoding='utf-8')
    # remove comments
    text = re.sub(r'//.*', '', text)
    text = re.sub(r'/\*.*?\*/', '', text, flags=re.S)
    # find first array or object literal assignment
    match = re.search(r'=\s*(\[.*\]);', text, re.S)
    if not match:
        raise ValueError('No array found in ' + path)
    js = match.group(1)
    # normalize unquoted keys and single quotes
    js = re.sub(r'([\{,\s])(\w+)\s*:', r'\1"\2":', js)
    js = re.sub(r"'([^']*)'", r'"\1"', js)
    js = re.sub(r',\s*([\]}])', r'\1', js)
    return json.loads(js)

barangays = parse_js_array('data/barangays.js')
fac = parse_js_array('data/facilities.js')

# build polygon map
polygons = {}
for b in barangays:
    coords = b['geojson']['geometry']['coordinates'][0]
    polygons[b['name']] = coords

# all facilities lists keys
lists = [v for k, v in fac.items() if isinstance(v, list)]
facilities = [item for sub in lists for item in sub]

from shapely.geometry import Point, Polygon

errors = []
for f in facilities:
    bar = f.get('barangay')
    if not bar: continue
    if bar not in polygons:
        errors.append((f['id'], f['name'], bar, 'barangay not found'))
        continue
    poly = Polygon(polygons[bar])
    pt = Point(f['lng'], f['lat'])
    if not poly.contains(pt) and not poly.touches(pt):
        errors.append((f['id'], f['name'], bar, f['lat'], f['lng']))

with open('temp_location_check.txt','w', encoding='utf-8') as out:
    if not errors:
        out.write('OK\n')
    else:
        for e in errors:
            out.write(str(e) + '\n')
print('done')
